lib/src/configuration.rs

```diff
@@ -0,0 +1,59 @@
+#[derive(Debug, Clone)]
+pub struct Configuration {
+    pub base_path: String,
+    pub user_agent: Option<String>,
+    pub basic_auth: Option<BasicAuth>,
+    pub oauth_access_token: Option<String>,
+    pub bearer_access_token: Option<String>,
+    pub api_key: Option<ApiKey>,
+}
+
+// Check whether defining BasicAuth works like this or not, else revert to the basic definition commented out
+#[derive(Debug, Clone)]
+pub struct BasicAuth {
+    pub username: String,
+    pub password: Option<String>,
+}
+// pub type BasicAuth = (String, Option<String>);
+
+#[derive(Debug, Clone)]
+pub struct ApiKey {
+    pub prefix: Option<String>,
+    pub key: String,
+}
+
+impl Configuration {
+    pub fn new(
+        base_path: String,
+        user_agent: Option<String>,
```

Comment:
There is no optional args in rust, only nullable one. So constructors with `Option<>` will require to provide all args anyway, e.g. 

```rust
Configuration::new("/path".to_string(), null, null, null)
```

It does not look really nice, but more important in the future if you want to add 1 more optional arguments to your configuration then **all previous use of the constructor will need an update to add one more `null`!**.

Either making your codebase too rigid, or forcing you to introduce breaking changes often

For future sustainability it is recommended to handle constructors with many optional args with the builder pattern: https://rust-unofficial.github.io/patterns/patterns/creational/builder.html

e.g. https://github.com/vemonet/nanopub-rs/blob/main/lib/src/profile.rs#L30

Reply:
Interesting, good to know :pray:

Reply:
Thanks for the review!
I didn't know about this, and I have now updated this PR to the builder pattern